### PR TITLE
Create new collection for Hcal lasermon digis + Hcal unpacker updates

### PR DESCRIPTION
**Pull request by jkunkle**
This pull request addresses three issues related to Hcal data unpacking.  (See more detailed explanations below) : 

1) Add a QIE10DigiCollection with tag "LASERMON" to keep laser monitoring data in the event
2) Improvements to the HcalUnpacker that are related to 1) 
3) Updates to RecoMET/METProducers/HcalNoiseInfoProducer to reflect the changes in 1)

Additional information to (1)
-------------------------------------
As we plan to use the lasermon channels to clean the data of any possible laser contamination, we desire to keep them in the emap so that we have a versioned accounting of them.  This requires changes to : 
- CalibFormats/HcalObjects/HcalTextToDetIdConverter, To parse the new emap entries from the .txt file
- DataFormats/HcalDetId/HcalCalibDetId, To add a unique CalibDetType
- EventFilter/HcalRawToDigi/HcalRawToDigi, to make the new collection and put it in the event
- EventFilter/HcalRawToDigi/HcalUnpacer, to unpack into the new collection

Additional information to (2)
-------------------------------------------
We often run into an issue with the HcalUnpacker where it encounters data having a different number of samples than the default collection (the number of samples is determined by the first FED of that type that is unpacked).  Currently the unpacker stops unpacking at this point and any data beyond that is not unpacked.  At this point in the code there are even comments that say "this is horrible".  The aim of this update is to avoid this situation and to provide access to this new data for private studies without making by-hand modifications to the unpacker.

This update does not change the default behavior of the unpacker except for 
1) The unpacker will no longer stop unpacking after encountering non-conforming data
2) the message that is printed in this case is modified to direct the user how to access the data

To access the data, the user can provide two lists to HcalRawToDigi, one that specifies the number of samples in the data, and a tag so the data can be stored in a unique collection.  There are two additional maps of integers (number of samples) to QIE10/11DataCollections.  The unpacker looks for these maps to be filled and unpacks accordingly.  Otherwise they follow the same processing as other QIE10/11 collections

Additional information to (3)
---------------------------------------
This just updates the code to read the lasermon channels from a QIE10DigiColleciton.  The algorithm does not change, there is just some structural reordering of the code to use the new collection

I have run the runTheMatrix tests and will soon provide instructions for testing the new features

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by jkunkle** [comment]
In discussing with colleagues and testing, Its clear that we needed to add a protection for the default qie10 and qie11 collections in the case that the data has a different number of samples.  This protection as assured in the past because the unpacker would return.  Here I just add the if statement that insures the protection

**Comment by jkunkle** [comment]
In testing I found that we weren't properly handling the data when a QIE reset was present.  I've added a protection by checking the ok() bit of the data frame

**Comment by perrotta** [comment]
please test

**Comment by deguio** [comment]
hey @jkunkle 
since the plan is to test laser operations during cosmics data taking, could you please backport this PR to 10_0_x?
thanks,
F.

**Comment by civanch** [comment]
+1

**Comment by jkunkle** [comment]
Some instructions for testing 

1) Reproduce original problem with unpacker encountering data of different sample length.  
-----------------------------------------------------------------------
cmsrel CMSSW_10_1_X_2018-02-28-2300 #or a reasonably new release
cd CMSSW_10_1_X_2018-02-28-2300/src
cmsenv
cp -r /afs/cern.ch/user/j/jkunkle/public/HcalUnpackerCodeTest/* .
scram b
cmsRun TestUnpackerCode/TestUnpackerCode/test/test_unpacker_cfg.py  inputFiles=file:/afs/cern.ch/work/j/jkunkle/public/CMS/HCAL/HFLaser/5E197134-2D17-E811-95ED-02163E011C3B.root processEvents=10
# should see the message QIE10ZDC Collection has 3 samples per digi, raw data has 6!

2) Get the updated code
-----------------------------------------------------------------------
cd $CMSSW_BASE/src
rm -rf * #otherwise git will refuse to checkout packages
git cms-merge-topic -u 22385
cp -r /afs/cern.ch/user/j/jkunkle/public/HcalUnpackerCodeTest/* .
scram b -j8
cmsRun TestUnpackerCode/TestUnpackerCode/test/test_unpacker_cfg.py  inputFiles=file:/afs/cern.ch/work/j/jkunkle/public/CMS/HCAL/HFLaser/5E197134-2D17-E811-95ED-02163E011C3B.root processEvents=10 getLasermonDigis=1
# Since the lasermon channels are being handled we get no complaint from the unpacker.   The test code also prints out that it could access the 24 lasermon digis

3) Produce unpacker message.  Use a local run where there is an additional readout having a different number of time slices
-----------------------------------------------------------------------------
cmsRun TestUnpackerCode/TestUnpackerCode/test/test_unpacker_cfg.py  inputFiles=root://eoscms.cern.ch//eos/cms/store/group/dpg_hcal/comm_hcal/USC/run310190/USC_310190.root processEvents=10 getLasermonDigis=1 local=1 
# now one should see the message "The default QIE10 Collection has 6 samples per digi, while the current data has 10! " with additional instructions on how to store the data having 10 samples

4) Read out additional data
---------------------------------------------------------------------------
cmsRun TestUnpackerCode/TestUnpackerCode/test/test_unpacker_cfg.py  inputFiles=root://eoscms.cern.ch//eos/cms/store/group/dpg_hcal/comm_hcal/USC/run310190/USC_310190.root processEvents=10 getLasermonDigis=1 local=1 getUMNIODigis=1
# this adds the recommended configurations from the error message in 3) and reads out the 4 channels in the test script

5) The following items test the cases where one may misconfigure the code.  Test the case when a tag is requested that already exists (ie LASERMON already exists)
-------------------------------------------------------------------------
cmsRun TestUnpackerCode/TestUnpackerCode/test/test_unpacker_cfg.py  inputFiles=root://eoscms.cern.ch//eos/cms/store/group/dpg_hcal/comm_hcal/USC/run310190/USC_310190.root processEvents=10 getLasermonDigis=1 local=1 testTagDuplication=1
# should get the message "Cannot create an additional QIE10 sample with name LASERMON, it is already created!  It must be removed along with the corresponding entry in saveQIE10DataNSamples" In this case only the default QIE10 collections are created

6) Test the case when the configured nSamples duplicates an existing value
------------------------------------------------------------------------
cmsRun TestUnpackerCode/TestUnpackerCode/test/test_unpacker_cfg.py  inputFiles=root://eoscms.cern.ch//eos/cms/store/group/dpg_hcal/comm_hcal/USC/run310190/USC_310190.root processEvents=10 getLasermonDigis=1 local=1 testNSamplesDuplication=1
#should see the message "QIE10 data requested to be stored in tag UMNIO is already stored in the default QIE10 collection.  To avoid duplicating, remove the tag UMNIO from the saveQIE10DataTags and the value of 6 from the saveQIE10DataNSamples configurables to HcalRawToDigi"

7) Test when the vector inputs are mismatched in length
-----------------------------------------------------------------------
cmsRun TestUnpackerCode/TestUnpackerCode/test/test_unpacker_cfg.py  inputFiles=root://eoscms.cern.ch//eos/cms/store/group/dpg_hcal/comm_hcal/USC/run310190/USC_310190.root processEvents=10 getLasermonDigis=1 local=1 testMismatch=1
# should see the message "The saveQIE10DataTags and saveQIE10DataNSamples inputs must be the same length!  These will be ingored"

**Comment by jkunkle** [comment]
For further testing, I just took a global run where we self trigger on the laser signal.   Data were recorded as in normal global runs, so we can check that the lasermon signal is propagated all the way to the MET filter.

I repeat the setup steps above (I added some additional code in my public area, so that should at least be copied again)
------------------------------------------------
cmsrel CMSSW_10_1_X_2018-02-28-2300 #or a reasonably new release
cd CMSSW_10_1_X_2018-02-28-2300/src
cmsenv
git cms-merge-topic -u 22385
cp -r /afs/cern.ch/user/j/jkunkle/public/HcalUnpackerCodeTest/* .
scram b -j8
cmsRun reco_RAW2DIGI_RECO.py 
cmsRun aod_PAT.py
cmsRun TestLaserCode/TestLaserCode/test/config_test.py 

There is a printout at the end of this code that checks that the METFilter bit corresponds correctly to the energy threshold that we want to apply.  It also creates histogram of the laser energy that can be found in lasermonHists.root

**Comment by jkunkle** [comment]
Hi all,  One thing that bothers me somewhat is that this code will give error messages on every event in the default configuration (this was true in the past as well, but it was avoided because we didn't read out the hardware).  If this is a concern I can add a simple message counter which only prints the first N times (which I would put at 20).  How does this sound?

**Comment by abdoulline** [comment]
Sounds reasonable (OK) to me...

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by jkunkle** [comment]
Hi All,

I have additionally tested by rerunning our online DQM to confirm that we can now have working monitoring (which was broken because the unpacker stops unpacking when it encounters data having differing number of samples).  I think that we in HCAL are happy with this update.  Can we aim to have it merged for the  10_1_0pre3 release?  Can I already make the pull request for backporting into 10_0_X?

File: EventFilter/HcalRawToDigi/plugins/HcalRawToDigi.cc [hunk 1]
```diff
@@ -194,25 +272,72 @@ void HcalRawToDigi::produce(edm::Event& e, const edm::EventSetup& es)
   }
   if( cntHFdup ) edm::LogError("HcalRawToDigi") << "Duplicated HF digis found for "<<cntHFdup<<" times"<<std::endl;
 
+  // Check that additional QIE10 and QIE11 collections
+  // do not duplicate the default collections
+  for( const auto & addtl: colls.qie10Addtl ) {
+      if( addtl.second->samples() == colls.qie10->samples() ) {
+          std::string tag = saveQIE10Info_[addtl.second->samples()];
+          edm::LogError("HcalRawToDigi") << "QIE10 data requested to be stored in tag "
```
Comment (perrotta) [inline EventFilter/HcalRawToDigi/plugins/HcalRawToDigi.cc]: Again: do these need to be LogError's?

File: EventFilter/HcalRawToDigi/src/HcalUnpacker.cc [hunk 2]
```diff
@@ -903,3 +935,35 @@ void HcalUnpacker::unpackUMNio(const FEDRawData& raw, int slot, Collections& col
...
 }
+
+void HcalUnpacker::printInvalidDataMessage( const std::string &coll_type, int default_ns, int conflict_ns, bool extended ) {
+
+  nPrinted_++;
+
+  int limit = 50;
+  if( nPrinted_ >= limit ) {
+
+      if( nPrinted_ == limit ) edm::LogError("Invalid Data") << "Suppresing further error messages" << std::endl;
```
Comment (perrotta) [inline EventFilter/HcalRawToDigi/src/HcalUnpacker.cc]: "Suppressing"...

File: EventFilter/HcalRawToDigi/src/HcalUnpacker.cc [hunk 3]
```diff
@@ -903,3 +935,35 @@ void HcalUnpacker::unpackUMNio(const FEDRawData& raw, int slot, Collections& col
   *(colls.umnio) = HcalUMNioDigi(data, nwords);
   
 }
+
+void HcalUnpacker::printInvalidDataMessage( const std::string &coll_type, int default_ns, int conflict_ns, bool extended ) {
+
+  nPrinted_++;
+
+  int limit = 50;
```
Comment (perrotta) [inline EventFilter/HcalRawToDigi/src/HcalUnpacker.cc]: Maybe the 20 that you suggested in the github thread can be enough...

File: EventFilter/HcalRawToDigi/src/HcalUnpacker.cc [hunk 4]
```diff
@@ -903,3 +935,35 @@ void HcalUnpacker::unpackUMNio(const FEDRawData& raw, int slot, Collections& col
...
+      message << "\nIn order to store this data in the event, it must have a unique tag.  "
+              << "To accomplish this, provide two lists to HcalRawToDigi \n"
+              << "1) that specifies the number of samples and "
+              << "2) that gives tags with which these data are saved.\n"
+              << "For example in this case you might add \n"
+              << "process.hcalDigis.save" << coll_type << "DataNSamples = cms.untracked.vint32( " 
+              << conflict_ns << ") \nprocess.hcalDigis.save" << coll_type << "DataTags = cms.untracked.vstring( \"MYDATA\" )" ;
+  }
+
+  edm::LogError("Invalid Data") << message.str() << std::endl;
```
Comment (perrotta) [inline EventFilter/HcalRawToDigi/src/HcalUnpacker.cc]: Again, is this needed to be a LogError?

**Comment by perrotta** [comment]
@jkunkle : it is not from your update, but the static analyzer is reporting a dead initialization at L601 of EventFilter/HcalRawToDigi/src/HcalUnpacker.cc

In fact, the ```nps``` initialized there is never used, it being re-initialized later at L617

Could you (or someone else from the HCAL group) please check if this witnesses some bug in the code? If not, L601 (and the previous comment at L600) could simply be removed, just to avoid misunderstanding about what happens "for the 1.6 Gbps link data"

**Comment by jkunkle** [comment]
Thanks, Andrea for the review.  I'll implement your suggestions (thanks for catching my spelling errors).  I've passed on the question about the issue with nps.  I'll wait until we have a resolution to that and push the changes together.

**Comment by jkunkle** [comment]
@kpedro88  it looks like you made this second nps calculation in 2016, https://github.com/cms-sw/cmssw/commit/ec08635309798b575434550d117683d0d1810b1e.  Do you expect that we can remove the old calculation of nps?

**Comment by kpedro88** [comment]
yes, I think we can remove the old one.

**Comment by perrotta** [comment]
please test

**Comment by civanch** [comment]
+1

File: RecoMET/METProducers/src/HcalNoiseInfoProducer.cc [hunk 5]
```diff
@@ -672,137 +649,180 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
...
+
+      HcalCalibDetId calibId( digi->id() );
+
+      // Fill the lasermonitor channels
+      int cboxch  = calibId.cboxChannel( );
+      int iphi    = calibId.iphi();
+      int ieta    = calibId.ieta();
+
+      // only check channels having the requested cboxch
+      if (std::find( laserMonCBoxList_.begin(), laserMonCBoxList_.end(),
```
Comment (perrotta) [inline RecoMET/METProducers/src/HcalNoiseInfoProducer.cc]: Is this ```find``` needed? If ```cboxch``` is found, then the loop over the same collection is re-done below; if it is not found, the check is similar as the ```if( cboxch == laserMonCBoxList_[idx])``` which is right at the beginning of the ```for``` loop below

File: EventFilter/HcalRawToDigi/src/HcalUnpacker.cc [hunk 6]
```diff
@@ -404,7 +405,7 @@ void HcalUnpacker::unpackVME(const FEDRawData& raw, const HcalElectronicsMap& em
 	  ncurr=0;
 	  valid=true;
 	}      
-	// add the word (if within settings or recent firmware [recent firmware ignores startSample/endSample])
+	// add the word (if within settings or recent firmware [recent firmware Suppressing startSample/endSample])
```
Comment (perrotta) [inline EventFilter/HcalRawToDigi/src/HcalUnpacker.cc]: "Suppressing"? (maybe inadvertently replaced by a copy/paste?)

**Review by perrotta — COMMENTED** [review]
With this I am done with the visual code check (I left the HcalNoiseInfoProducer behind last time).

Tomorrow, I will test this PR following your recipe

File: RecoMET/METProducers/src/HcalNoiseInfoProducer.cc (hunk 5 above)
Comment (jkunkle) [reply RecoMET/METProducers/src/HcalNoiseInfoProducer.cc]: yes, we discussed this the first time I committed this code and it didn't get removed.  I'll remove it.

File: EventFilter/HcalRawToDigi/src/HcalUnpacker.cc (hunk 6 above)
Comment (jkunkle) [reply EventFilter/HcalRawToDigi/src/HcalUnpacker.cc]: must have been, I'll revert it

**Comment by perrotta** [comment]
please test

**Comment by jkunkle** [comment]
I no longer see my prescription for running tests in the thread here, is it just me?

**Comment by abdoulline** [comment]
Me neither...
There is a box "67 hidden items, Load more..."
at position  between 8 days ago and 3 day ago, but 
clicking on it doesn't bring anything...

**Comment by perrotta** [comment]
The same happens to me: it looks like github does not dsplay any more all the messages in the thread hidden behind the button
immagine

However, I can still access them through the direct link that I saved:
https://github.com/cms-sw/cmssw/pull/22385#issuecomment-369558236

jkunkle <notifications@github.com> ha scritto:

**Comment by jkunkle** [comment]
I ran the testing myself and found that the first step didn't work.  This is because I changed the emap, I've updated the testing code to read in the original emap for the first step

**Comment by perrotta** [comment]
+1
- Keep laser monitoring data in the event
- In case of data with extra samples in FED, those extra samples are now also unpacked and saved for further private studies (not being used for the current workflows)
- The chain of savings and warning messages designed for it had been tested with the test program provided (not in this PR)
- No change expected and observed in reco outputs in the standard workflows, while new output will get stored in case of lasermon events or extra samples in the FED

**Comment by mariadalfonso** [comment]
@jkunkle 
now that is signed by RECO  to you want to make the backport in the 10_0 ?

We should have the backport in the 10_0_4 since we want to test these changes at P5 asap.

Thanks

**Comment by fabiocos** [comment]
@lpernie @arunhep @ghellwig @civanch could you please check this PR?

**Comment by civanch** [comment]
+1

**Comment by lpernie** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Pull request merged**